### ISSUE TITLE
consider refactor to scikit-style mixins

### ISSUE THREAD

**Issue by knaaptime** [pysal/segregation#4]
@renanxcortes and I have had some conversations about refactoring this project so that, instead of each segregation index being implemented as its own class, we might have two classes (e.g. spatial and aspatial), and the indices themselves would move to subclasses or functions.

In another thread, as we work on [spopt](https://github.com/pysal/spopt) we're making a concerted effort to use the BaseClass/Mixin structure that scikit-learn uses. I want to raise the option here of adopting a similar architecture, which i think would make a lot of sense for this project

**Comment by renanxcortes** [comment]
Yes, I totally agree that refactoring like that would be much better! This is something that I'll start shortly in a branch!

**Comment by renanxcortes** [comment]
@knaaptime I recently update this project to this new style! I'm still working on some other improvements, but I think we can close this issue, right?

**Comment by knaaptime** [comment]
Yup,  nice work

**Comment by renanxcortes** [comment]
I'm reopening this issue, since the current version is an improvement of the subclasses with the type:

index_type.index_type

But not as the Mixins type mentioned in the first comment that scikit-learn does.  Such as:

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/affinity_propagation_.py

**Comment by knaaptime** [comment]
I'm eyeballing this refactor sometime in the near future. What i'd like to do is target this as segregation 2.0 because the api will change pretty substantially. What I have planned is to do away with the spatial/aspatial distinction in the module and make it clear that [almost] any of the indices can be spatialized by passing a `W` or a `Network` to the init. 

A more meaningful distinction will come in the form of two base classes: `SingleGroupIndex` and `MultiGroupIndex`, each of which can be spatial or aspatial. Currently this is a muddy division in the library, with "aspatial" indices divided into single and multigroup modules and spatial indices all lumped together. 

the new modules will be 

**indices**:
`singlegroup`
`multigroup`
`local`

**analysis**
`decomposition`
`inference`

**Utility** (subsumes `network` and `compute_all`)
`util`

**Comment by knaaptime** [comment]
```python
from segregation.singlegroup import GiniSeg
from segregation.multigroup import MultiGiniSeg

g_a = GiniSeg(group_pop_var, total_pop_var)   # aspatial single group
g_s = GiniSeg(group_pop_var, total_pop_var, W)  # spatial single group

mg_a = MultiGiniSeg(input_df, groups_list)  # aspatial multigroup
mg_s = MultiGiniSeg(input_df, groups_list, Network)  # spatial multigroup
```

**Comment by knaaptime** [comment]
this will make it *much* easier to accomodate #104 and will also make the decomposition easier to handle for spatial indices

**Comment by knaaptime** [comment]
i've started this in my 2.0 branch

**Comment by knaaptime** [comment]
resolved by #161
